clippy_lints/src/assertions_on_constants.rs

```diff
@@ -63,7 +65,93 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for AssertionsOnConstants {
+                        e.span,
+                        "`assert!(false)` should probably be replaced",
+                        "use `panic!()` or `unreachable!()`",
+                    );
+                } else {
+                    span_help_and_lint(
+                        cx,
+                        ASSERTIONS_ON_CONSTANTS,
+                        e.span,
+                        &format!("`assert!(false, \"{}\")` should probably be replaced", panic_message,),
+                        &format!(
+                            "use `panic!(\"{}\")` or `unreachable!(\"{}\")`",
+                            panic_message, panic_message,
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+// fn get_assert_args(snip: String) -> Option<Vec<String>> {
+//
+// }
+
+fn assert_with_message<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) -> Option<(LocalInternedString, bool)> {
+    if_chain! {
+        if let ExprKind::Match(ref expr, ref arms, MatchSource::IfDesugar { contains_else_clause: false }) = expr.kind;
+        // match expr
+        if let ExprKind::DropTemps(ref expr) = expr.kind;
+        if let ExprKind::Unary(UnOp::UnNot, ref expr) = expr.kind;
+        //if let ExprKind::Lit(ref lit) = expr.kind;
+        if let Some((Constant::Bool(is_true), _)) = constant(cx, cx.tables, expr);
+        //if is_true;
+        // match arms
+        // arm 1 pattern
+        if let PatKind::Lit(ref lit_expr) = arms[0].pat.kind;
+        if let ExprKind::Lit(ref lit) = lit_expr.kind;
+        if let LitKind::Bool(true) = lit.node;
+        //if let LitKind::Bool(true) = lit1.node;
+        // arm 1 block
+        if let ExprKind::Block(ref block1, _) = arms[0].body.kind;
+        if let Some(trailing_expr1) = &block1.expr;
+        if block1.stmts.len() == 0;
+        //
+        if let ExprKind::Block(ref actual_block1, _) = trailing_expr1.kind;
+        if let Some(block1_expr) = &actual_block1.expr;
+        // function call
+        if let ExprKind::Call(ref func, ref args) = block1_expr.kind;
+        if let ExprKind::Path(ref path) = func.kind;
+        // ["{{root}}", "std", "rt", "begin_panic"] does not work
+        if match_qpath(path, &["$crate", "rt", "begin_panic"]);
+        // arguments
+        if args.len() == 2;
+        if let ExprKind::Lit(ref lit) = args[0].kind;
+        if let LitKind::Str(ref s, _) = lit.node;
+        let panic_message = s.as_str(); // bind the panic message
+        if let ExprKind::AddrOf(MutImmutable, ref inner) = args[1].kind;
+        if let ExprKind::Tup(ref elements) = inner.kind;
+        if elements.len() == 3;
+        if let ExprKind::Lit(ref lit1) = elements[0].kind;
+        if let LitKind::Str(ref s1, _) = lit1.node;
+        if let ExprKind::Lit(ref lit2) = elements[1].kind;
+        if let LitKind::Int(_, _) = lit2.node;
+        if let ExprKind::Lit(ref lit3) = elements[2].kind;
+        if let LitKind::Int(_, _) = lit3.node;
```

Comment:
I don't think that you have to match the second argument of the `begin_panic` call.